### PR TITLE
JP Connect: Handle User Already Connected Case in Redux Actions

### PR DESCRIPTION
**Pull request by ockham**
This PR implements the following rough plan of action (that could be turned into a stretegy for other JPC auth notices):

* Start in `jetpack-connect-notices`: 
  * Locate `USER_IS_ALREADY_CONNECTED_TO_SITE` case, remove.
  * Locate what selector was used to determine that case: `getUserAlreadyConnected`. Remove from `connect`.
* Remove that selector and its unit test.
* Make note what part of the state this was using: `state.jetpackConnect.jetpackConnectAuthorize.userAlreadyConnected`. Locate corresponding [reducer](https://github.com/Automattic/wp-calypso/blob/b1ec73be87889566e96b501171ff00004fee8c93/client/state/jetpack-connect/reducer/jetpack-connect-authorize.js#L76-L77). 
* Remove reducer and its corresponding unit test.
* Make note what action type name the reducer was listening to (`JETPACK_CONNECT_USER_ALREADY_CONNECTED`), locate corresponding action.
* Have notices middleware listen to action. If on mobile, dispatch a redirect action.

### Testing Instructions

See @sirreal's [comment](https://github.com/Automattic/wp-calypso/pull/24645#issuecomment-386254827) and the screenshots below. (You have to add `&calypso_env=development` to the JP Connect URL to test this branch.)

### References
(showing how the mobile redirect was implemented in `jetpack-connect/authorize`): 
https://github.com/Automattic/wp-calypso/blob/b1ec73be87889566e96b501171ff00004fee8c93/client/jetpack-connect/authorize.js#L376-L382

https://github.com/Automattic/wp-calypso/blob/b1ec73be87889566e96b501171ff00004fee8c93/client/jetpack-connect/authorize.js#L653-L654

### Screenshots

Before:

image

After:

image

### CODE REVIEW

File: client/state/notices/middleware.js [hunk 1]
```diff
@@ -319,6 +320,15 @@ export const onBillingTransactionRequestFailure = ( { transactionId, error } ) =
...
 
+export const onJetpackConnectUserAlreadyConnected = () => {
+	return errorNotice(
+		translate(
+			'This WordPress.com account is already connected to another user on this site. ' +
+				'Please login to another WordPress.com account to complete the connection.'
+		)
+	);
+};
+
```
Comment (ockham) [inline client/state/notices/middleware.js:326]: Or should we make this a generic `onJetpackConnectFailure = message => {...}` handler?

**Comment by ockham** [comment]
I should probably add some unit tests to cover the new/moved functionality...

**Comment by sirreal** [comment]
Instructions to reproduce:

1. Connect a site
1. Add a user with privileges to connect/manage (administrator)
1. Log out and log in with the new user
1. Connect as that user via WP Admin (/wp-admin/admin.php?page=jetpack#/dashboard) while still logged in to WordPress.com with the same acct:
   connect
1. You'll hit that error:
   already-connected


To test mobile, first visit `/jetpack/connect/?mobile_redirect=wordpress://`, that will set the mobile cookie, then try visiting the auth URL again. Remember to clear the cookie when you're finished.

File: client/state/action-types.js [hunk 2]
```diff
@@ -366,7 +366,7 @@ export const JETPACK_CONNECT_SSO_VALIDATION_ERROR = 'JETPACK_CONNECT_SSO_VALIDAT
 export const JETPACK_CONNECT_SSO_VALIDATION_REQUEST = 'JETPACK_CONNECT_SSO_VALIDATION_REQUEST';
 export const JETPACK_CONNECT_SSO_VALIDATION_SUCCESS = 'JETPACK_CONNECT_SSO_VALIDATION_SUCCESS';
 export const JETPACK_CONNECT_STORE_SESSION = 'JETPACK_CONNECT_STORE_SESSION';
-export const JETPACK_CONNECT_USER_ALREADY_CONNECTED = 'JETPACK_CONNECT_USER_ALREADY_CONNECTED';
+export const JETPACK_CONNECT_AUTHORIZE_FAILURE = 'JETPACK_CONNECT_AUTHORIZE_FAILURE';
```
Comment (tyxla) [inline client/state/action-types.js]: We should move that action type a bit up 🔤

File: client/state/notices/middleware.js [hunk 3]
```diff
@@ -319,6 +320,10 @@ export const onBillingTransactionRequestFailure = ( { transactionId, error } ) =
 	} );
 };
 
+export const onJetpackConnectAuthorizeFailure = ( { message } ) => {
```
Comment (tyxla) [inline client/state/notices/middleware.js]:
```
( { message } ) => {
	return errorNotice( message );
}
```

could be shortened to 

```
( { message } ) => errorNotice( message );
```

File: client/state/notices/middleware.js [hunk 4]
```diff
@@ -319,6 +320,10 @@ export const onBillingTransactionRequestFailure = ( { transactionId, error } ) =
...
+	return errorNotice( message );
```
Comment (tyxla) [inline client/state/notices/middleware.js]: The previous notice was a little less scary (yellow). Should this one be a `warningNotice` instead? Or perhaps we could have a `type` in addition to the `message` that would specify the notice type?

File: client/state/jetpack-connect/actions.js [hunk 5]
```diff
@@ -276,6 +279,16 @@ export function createAccount( userData ) {
 	};
 }
 
+export function redirectMobileApp( reason ) {
```
Comment (tyxla) [inline client/state/jetpack-connect/actions.js]: Perhaps add some tests for this one?

**Review by tyxla — COMMENTED** [review]
This works well in my testing 👍  Code also looks good. 

Our action creators could use some better testing coverage. Also, I've left some other minor suggestions, nothing critical.

**Comment by tyxla** [comment]
I think this redufixication could still be beneficial for code simplification, it's just out of our priority list for now, as we've switched focus to other projects.

**Comment by sirreal** [comment]
What do you think folks, shall we let this one go?

**Comment by tyxla** [comment]
> What do you think folks, shall we let this one go?

I think so unless our friends from @Automattic/jetpack-voyager would like to pick it up where we left off.

**Comment by sirreal** [comment]
Closing. Anyone who wants to take over can reopen.

**Pull request closed without merging**